### DISCUSSION TITLE
WSL  $PATH dir spaces issue

### DISCUSSION THREAD

**Discussion by nicriz** [microsoft/WSL#6311] [General] [answered]
# Environment

```none
Windows build number: [Version 10.0.18363.1198]
Your Distribution version: [Ubuntu 20.04]
Whether the issue is on WSL 2 and/or WSL 1: [Linux version 4.4.0-18362-Microsoft (Microsoft@Microsoft.com) (gcc version 5.4.0 (GCC) ) #1049-Microsoft Thu Aug 14 12:01:00 PST 2020]
```

# Steps to reproduce
Executing this script.
```
export tmpnccfg="$( cd -P "$( dirname "${BASH_SOURCE[0]}" )" && pwd )/bin/ncrystal-config"
if [ -f "$tmpnccfg" ]; then
    $( "$tmpnccfg" --setup )
    unset tmpnccfg
    return 0
else
    echo "ERROR: Could not find ncrystal-config script"
    unset tmpnccfg
    return 1
fi
```
The -config file executed by the script is [here](https://gist.github.com/nicriz/ba2170164bec4fca385ae8a94a29d791).
#  Expected behavior

I expected to set up the environmental variables for the software.

# Actual behavior
```
-bash: export: `'PATH=/home/nicorizzi/ncrystal_install/bin:/home/nicorizzi/phits/dchain-sp/bin:/home/nicorizzi/phits/bin:/home/nicorizzi/anaconda3/bin:/home/nicorizzi/anaconda3/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/mnt/c/Python27/Lib/site-packages/PyQt4:/mnt/c/Program': not a valid identifier
-bash: export: `Files/Python38/Scripts:/mnt/c/Program': not a valid identifier
-bash: export: `Files/Python38:/mnt/c/WINDOWS/system32:/mnt/c/WINDOWS:/mnt/c/WINDOWS/System32/Wbem:/mnt/c/WINDOWS/System32/WindowsPowerShell/v1.0:/mnt/c/WINDOWS/System32/OpenSSH:/mnt/c/Python27:/mnt/c/Python27/DLLs:/mnt/c/Python27/Scripts:/mnt/c/Python27/Lib/site-packages/vtk:/mnt/c/Python27/gnuplot/binary:/mnt/c/Program': not a valid identifier
-bash: export: `(x86)/pythonxy/SciTE-3.2.2-1:/mnt/c/Program': not a valid identifier
-bash: export: `(x86)/pythonxy/console:/mnt/c/MinGW32-xy/bin:/mnt/c/Program': not a valid identifier
-bash: export: `(x86)/MiKTeX': not a valid identifier
-bash: export: `2.9/miktex/bin:/mnt/c/Users/nicri/AppData/Local/Microsoft/WindowsApps:/mnt/c/Users/nicri/AppData/Local/Programs/Microsoft': not a valid identifier
-bash: export: `Code/bin:/snap/bin'': not a valid identifier
```
# Long Story

Hi everybody, I have a problem with the environmental variable PATH on WSL. I am trying to install a software called [NCrystal ](https://github.com/mctools/ncrystal), which requires to set up some environmental variables. However, when I try to execute the script given by the developers I end up with the error above. 

It seems like Linux is breaking the PATH variable due to the spaces in windows directories.
I found many issues on the topic, but I could not find a solution. Disabling the Windows variables on my WSL environment would not be a desirable solution for me, since I often use VS Code and I like the fact that I can run it from the terminal. 

Here you can find the PATH variable as printed out by `env` command:
```
PATH=/home/nicorizzi/phits/dchain-sp/bin:/home/nicorizzi/phits/bin:/home/nicorizzi/anaconda3/bin:/home/nicorizzi/anaconda3/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/mnt/c/Python27/Lib/site-packages/PyQt4:/mnt/c/Program Files/Python38/Scripts:/mnt/c/Program Files/Python38:/mnt/c/WINDOWS/system32:/mnt/c/WINDOWS:/mnt/c/WINDOWS/System32/Wbem:/mnt/c/WINDOWS/System32/WindowsPowerShell/v1.0:/mnt/c/WINDOWS/System32/OpenSSH:/mnt/c/Python27:/mnt/c/Python27/DLLs:/mnt/c/Python27/Scripts:/mnt/c/Python27/Lib/site-packages/vtk:/mnt/c/Python27/gnuplot/binary:/mnt/c/Program Files (x86)/pythonxy/SciTE-3.2.2-1:/mnt/c/Program Files (x86)/pythonxy/console:/mnt/c/MinGW32-xy/bin:/mnt/c/Program Files (x86)/MiKTeX 2.9/miktex/bin:/mnt/c/Users/nicri/AppData/Local/Microsoft/WindowsApps:/mnt/c/Users/nicri/AppData/Local/Programs/Microsoft VS Code/bin:/snap/bin
```

Thank you for your time

**Comment by DHowett** [comment]
Sorry, you’re trying to _run_ the contents of the environment variable `PATH`? You’re getting a “no such file or directory” error because `/bin:/sbin` (example) isn’t a fully qualified path to a single file that exists on your filesystem, as would be required to execute.

**Comment by nicriz** [comment]
> Sorry, you’re trying to _run_ the contents of the environment variable `PATH`? You’re getting a “no such file or directory” error because `/bin:/sbin` (example) isn’t a fully qualified path to a single file that exists on your filesystem, as would be required to execute.

Apologies, I thought it was the command to show the saved PATH, not execute it. I can change the description of the issue, but the problem when trying to export the environmental variable persists though

**Comment by therealkenc** [comment, marked as answer]
This is linux-behavior, variation #1766 #5099, others. Best guess the error in the script is not the `export` in the edited OP, but further up (wherever `$PATH` gets parsed). [Unintuitively enough, that `export tmpnccfg` looks okay, at least from here. The error described is `not a valid identifier`, and `tmpnccfg` is a valid identifier.]

If you bring it up in NCrystal-related forum, it is usually better to frame the ask in Linux terms instead of WSL. Looks like:
```
$ export MY_BIN="$HOME/My Things/bin"
$ mkdir -p "$MY_BIN"
$ export PATH=$PATH:$MY_BIN
$ # do NCrystal things
```

**Reply by nicriz** [reply]
Thank you, I'll try to give a careful look into the config file. Unfortunately the NCrystal developer does not have the workforce to support windows user :/
